utils/frozen_sstring.hh

```diff
@@ -0,0 +1,190 @@
+    size_t hash() const noexcept { return _hash; }
+
+    bool empty() const noexcept { return str().empty(); }
+    size_t size() const noexcept { return str().size(); }
+
+    const_iterator begin() const noexcept { return str().begin(); }
+    const_iterator end() const noexcept { return str().end(); }
+    const_iterator cbegin() const noexcept { return str().cbegin(); }
+    const_iterator cend() const noexcept { return str().cend(); }
+
+    const_reference front() const noexcept { return str().front(); }
+    const_reference back() const noexcept { return str().back(); }
+
+    const char_type& at(size_t pos) const { return str().at(pos); }
+
+    constexpr size_t find(const char_type* s, size_t pos = 0) const noexcept {
+        return str().find(s, pos);
+    }
+    size_t find(char_type t, size_t pos = 0) const noexcept {
+        return str().find(t, pos);
+    }
+    size_t find(const char_type* c_str, size_t pos, size_t len2) const noexcept {
+        return str().find(c_str, pos, len2);
+    }
+    size_t find(const basic_sstring& s, size_t pos = 0) const noexcept {
+        return find(s.c_str(), pos, s.size());
+    }
+    size_t find(const basic_frozen_sstring& s, size_t pos = 0) const noexcept {
+        return find(s.c_str(), pos, s.size());
+    }
+    template<class StringViewLike,
+             std::enable_if_t<std::is_convertible_v<StringViewLike,
+                                                    std::basic_string_view<char_type, traits_type>>,
+                              int> = 0>
+    size_t find(const StringViewLike& sv_like, size_type pos = 0) const noexcept {
+        return str().find(sv_like, pos);
+    }
+    size_t find_last_of(char_type c, size_t pos = npos) const noexcept {
+        return str().find_last_of(c, pos);
+    }
+
+    basic_frozen_sstring substr(size_t from, size_t len = npos) const {
+        return str().substr(from, len);
+    }
+
+    constexpr bool starts_with(std::basic_string_view<char_type, traits_type> sv) const noexcept {
+        return str().starts_with(sv);
+    }
+    constexpr bool starts_with(char_type c) const noexcept {
+        return str().starts_with(c);
+    }
+    constexpr bool starts_with(const char_type* s) const noexcept {
+        return str().starts_with(s);
+    }
+    constexpr bool ends_with(std::basic_string_view<char_type, traits_type> sv) const noexcept {
+        return str().ends_with(sv);
+    }
+    constexpr bool ends_with(char_type c) const noexcept {
+        return str().ends_with(c);
+    }
+    constexpr bool ends_with(const char_type* s) const noexcept {
+        return str().ends_with(s);
+    }
+    constexpr bool contains(std::basic_string_view<char_type, traits_type> sv) const noexcept {
+        return str().contains(sv);
+    }
+    constexpr bool contains(char_type c) const noexcept {
+        return str().contains(c);
+    }
+    constexpr bool contains(const char_type* s) const noexcept {
+        return str().contains(s);
+    }
```

Comment:
Do we really need to add proxies for all these? I think we can get away with just `operator==()` and a `const sstring& str()` accessor which gives access to the underlying string, if anybody wants to do operations on it.

Reply:
dropped in v2